### PR TITLE
add SDL backend based on sdl2 crate

### PR DESCRIPTION
**Pull request by matikij**
This PR adds a sdl backend

### CODE REVIEW

**Comment by djmuted** [comment]
I tested it and it works much better with this code in write function:
`    fn write(&mut self, data: &[i16]) -> io::Result<()> {
        while self.queue.size() > (2 * 2 * 44100) {
            // sleep and wait for sdl thread to drain the queue a bit
            thread::sleep(time::Duration::from_millis(10));
        }
        self.queue.queue(data);
        Ok(())
    }`
Your parameters for checking current queue size were a bit too small, making very short pauses in playback.

**Comment by matikij** [comment]
I have another version using `mpsc::sync_channel` (pushed it a couple of secs before), it uses way less CPU on my box, though the buffer size might be a bit small (100 ms). Could you test this and see if choppy playback issue is not present?

**Comment by djmuted** [comment]
The new version works fine, using ~9% of my CPU (i5-4690k 4.5GHz), while the previous (fixed) version uses 1-2%. I don't know why it uses less CPU on your box, but I'll stick to the first version.

**Comment by matikij** [comment]
Hmm, redid my testing, and numbers you reported are more-or-less similar here. Not sure where I got  the impression that the mpsc::sync_channel version uses less CPU, maybe I benchmarked debug builds. Anyway I used the parameters you have provided (and merged all commits into one).

**Comment by sashahilton00** [comment]
Hi guys, will take a look at this PR once #277 is dealt with.

**Comment by sashahilton00** [comment]
#277 was just merged, we ended up not changing the audio backend architecture just yet. Hence will review and merge this, though it will be removed with other backends once daemon is written.

File: playback/src/audio_backend/sdl.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+pub struct SdlSink {
+    queue: AudioQueue<Channel>,
+}
+
+impl Open for SdlSink {
+    fn open(device: Option<String>) -> SdlSink {
+        debug!("Using SDL sink");
+
+        if device.is_some() {
+            panic!("SDL sink does not support specifyng a device name");
```
Comment (sashahilton00) [inline playback/src/audio_backend/sdl.rs]: Typo on `specifying`

File: playback/src/audio_backend/sdl.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        }
+
+        let ctx = sdl2::init().expect("Could not init SDL");
+        let audio = ctx.audio().expect("Could not init SDL audio subsystem");
+
+        let desired_spec = AudioSpecDesired {
+            freq: Some(44_100),
+            channels: Some(2),
+            samples: None, // default
```
Comment (sashahilton00) [inline playback/src/audio_backend/sdl.rs]: annotation not needed.

File: playback/src/audio_backend/sdl.rs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            .open_queue(None, &desired_spec)
+            .expect("Could not open SDL audio device");
+
+        SdlSink { queue: queue }
+    }
+}
+
+impl Sink for SdlSink {
+    fn start(&mut self) -> io::Result<()> {
+        self.queue.clear(); //
```
Comment (sashahilton00) [inline playback/src/audio_backend/sdl.rs]: Please remove trailing slashes.

File: playback/src/audio_backend/sdl.rs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    fn stop(&mut self) -> io::Result<()> {
+        self.queue.pause();
+        self.queue.clear();
+        Ok(())
+    }
+
+    fn write(&mut self, data: &[i16]) -> io::Result<()> {
+        while self.queue.size() > (2 * 2 * 44_100) { 
+            // sleep and wait for sdl thread to drain the queue a bit
```
Comment (sashahilton00) [inline playback/src/audio_backend/sdl.rs]: Keep this annotation for future reference.

Comment (matikij) [reply playback/src/audio_backend/sdl.rs]: Not sure what I should here?

**Pull request merged**